config/abilities/provider_any.rb

```diff
@@ -44,7 +44,7 @@
 
     if user.has_permission?(:monitoring)
       can [:show], AlertRelatedEvent do |event|
-        user.has_access_to_service?(event.try(:service))
+        user.has_access_to_service?(event.try(:service) || event.service_id)
```

Comment:
This is to cover a situation where service was removed? Although I wonder if there is a test for this because I have some recollection that we were doing additional filtering for the return values of what services a user can access. But last person to touch that I think was @mayorova so maybe she would have more insight.

Reply:
This change is necessary because now the events have that `service_id` attribute, check this:

https://github.com/3scale/porta/blob/cd71af6c5cf46364f779ee94daa6b3f1326a575a/app/events/alerts/limit_reached_creatable.rb#L8

About that filtering you mention, I think it was me, not Daria, you mean this right? https://github.com/3scale/porta/pull/4029

That change in this line is necessary precisely to make that filter work after I changed the event above.

Anyway, I think it should be safer to do it like this:

```suggestion
        user.has_access_to_service?(event.try(:service) || event.try(:service_id))
```

Gonna change that.

Reply:
I think this is what I meant https://github.com/3scale/porta/pull/3929

But basically I'd like to see tests for this. If the service was removed, I expect that `has_access_to_service?` will return `false`. `try` should not be needed IMO.

Reply:
If I understand this correctly, you mean what happens when an event is referring to a service stored in its data, but then at the moment of executing the job, that service doesn't exist any more, right?

In that case, that's up to `has_access_to_service?`, I don't think that's related to this PR. Anyway, I did some tests manually. When the user has access to all services, then `has_access_to_service?` will return `true` when you pass it a service that doesn't exist. When the user has access to no services or to a list of particular services, `has_access_to_service?` will return `false` when passing it a non-existing service.

The behavior is not consistent and could lead to wrong filtering, but as I said I think that's out of this PR scope.

Reply:
Yeah, when the user has access to particular services and service does not exist, they will not be notified. That's the problematic use case. The cases where user has access to all or no services are correct as far as I understand.

IIRC this is because `has_access_to_service?` filters the services in a sense they have to exist and be owned by the provider and not sure if there was something else.

I find this a great safety mechanism and we probably have a test for it.

Probably we would need a new method like `had_access_to_service?` or idk.

So I'm fine to leave it as is + a JIRA for later investigation because presently we have a net improvement of the functionality.